### PR TITLE
urcrypt: integrate

### PR DESCRIPTION
**Pull request by joemfb**
This PR integrates `urcrypt`: @frodwith's new urbit crypto library. This project encapsulates urbit's standard crypto dependencies (openssl, secp256k1, ed25519, scrypt, libaes_siv) and custom variants (argon2u, ge-additions) into a singular library, suitable for use in other projects. For more details, see the [urcrypt README](https://github.com/urbit/urbit/blob/f5676a9b0fe77d1bd863a5d47ac2633abe6ccec0/pkg/urcrypt/README.md).

The development of urcrypt necessarily entailed significant refactoring and of vere's crypto jets; notable benefits include normalized logging and error handling, with new u3 apis for safe noun deconstruction. Many bugs were fixed, missing jet configuration was rectified (scrypt), and the secp256k1 implementation was significantly improved in both the jet and the hoon. All credit to @frodwith -- I am but a humble integrator.

------

Since `urcrypt` was developed on a branch predating #3806, and this PR is targeting a release-candidate branch that includes #4675, most of my integration work has revolved around the build system. Statically-linking vere with urcrypt involves resolving transitive library dependencies, which lead me to revise our use of `pkg-config`.

- @brendanhay, please review the `nix` and `pkg/urbit/configure` changes for sanity
- @locpyl-tidnyd, please take a look at `nix/sources*`, `pkg/urbit/compat/` and `pkg/urbit/configure`
- @frodwith, please review for completeness and intent

I owe a review of the `u3` changes.

------

TODO:

- [x] this branch includes binaries in its history (`ed25519.dll`, non-lfs pill)
- [x] `pkg/urcrypt` needs an autotools appropriate .gitignore (perhaps https://github.com/github/gitignore/blob/master/Autotools.gitignore)
- [ ] urcrypt could use a linkable version string/symbol (for `urbit -R`)
- [x] consider re-including urcrypt's vendored dependencies with `git sub-tree`
- [ ] the mingw build downloads the ivory pill as html (githubs raw download behavior seems to have a recently changed)
- [ ] the mingw build cache key for local builds isn't precise enough (doesn't affect CI, maybe out of scope)

### CODE REVIEW

**Comment by locpyl-tidnyd** [comment]
Re mingw/brew pill download, perhaps in your environment `$URBIT_VERSION` is set to a version that does not yet have a corresponding branch in `urbit/urbit` and curl is getting 404? I've just tried `1.3` and `1.5` and curl downloads the binary pill rather than some html.

**Comment by joemfb** [comment]
This PR has now been split into three, with dependency prep in #5170, the raw merge in #5171 (broken), and all build fixes, review changes, cleanup, &c remaining here.

File: nix/pkgs/urbit/default.nix [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
-, doCheck ? true, enableParallelBuilding ? true, dontStrip ? true }:
+{ lib, stdenv, coreutils, pkgconfig                      # build/env
+, cacert, ca-bundle, ivory                               # codegen
+, curlMinimal, ent, gmp, h2o, libsigsegv, libuv, lmdb    # libs
+, murmur3, openssl, softfloat3, urcrypt, zlib            #
+, enableStatic           ? stdenv.hostPlatform.isStatic  # opts
+, enableDebug            ? false
+, doCheck                ? true
+, enableParallelBuilding ? true
+, dontStrip              ? true }:
```
Comment (brendanhay) [inline nix/pkgs/urbit/default.nix:9]: Note: the prior formatting is a result of `sh/fmt-nix` (`nixfmt`).

**Review by brendanhay — APPROVED** [review]
Looks good - nothing stands out as cause to ready the guillotine.

**Pull request merged**